File: src/common/py/features/project_feature.py
import abc
from enum import IntEnum

from ..data.entities import ProjectFeatureID


class ProjectFeature(abc.ABC):
    """
    A project feature containing meta information describing a single project feature.
    """

    class FeatureFlags(IntEnum):
        """
        Possible flags for a project feature.
        """

        NONE = 0

        OPTIONAL = 1 << 0

    def __init__(
        self, *, feature_id: ProjectFeatureID, flags: FeatureFlags, display_name: str
    ):
        self._feature_id = feature_id

        self._flags = flags

        self._display_name = display_name

    @property
    def feature_id(self) -> ProjectFeatureID:
        return self._feature_id

    def has_flags(self, flags: FeatureFlags) -> bool:
        return (self._flags & flags) == flags

    @property
    def display_name(self) -> str:
        return self._display_name

File: src/common/web/features/ProjectFeature.ts
import { ProjectFeatureID } from "../data/entities/EntityTypes";

/**
 * Possible flags for a project feature.
 */
export const enum ProjectFeatureFlags {
    None = 0,

    Optional = 1 << 0
}

/**
 * A project feature containing meta information describing a single project feature.
 */
export abstract class ProjectFeature {
    private readonly _featureID: ProjectFeatureID;

    private readonly _flags: ProjectFeatureFlags;

    private readonly _displayName: string;
    private readonly _optionName: string;

    /**
     * @param featureID - The project feature ID.
     * @param flags - Feature flags.
     * @param displayName - The general display name.
     * @param optionName - The toggle-option name (only effective for optional features).
     */
    protected constructor(featureID: ProjectFeatureID, flags: ProjectFeatureFlags, displayName: string, optionName: string = "") {
        this._featureID = featureID;

        this._flags = flags;

        this._displayName = displayName;
        this._optionName = optionName;
    }

    /**
     * The project feature ID.
     */
    public get featureID(): ProjectFeatureID {
        return this._featureID;
    }

    /**
     * Check if certain flags are set.
     */
    public hasFlags(flags: ProjectFeatureFlags): boolean {
        return (this._flags & flags) === flags;
    }

    /**
     * The general display name.
     */
    public get displayName(): string {
        return this._displayName;
    }

    /**
     * he toggle-option name.
     */
    public get optionName(): string {
        return this._optionName;
    }
}
